Update our container support

eric-anderson opened the pull request:
to a more general approach that can work well with multiple dependent containers
fix a bunch of issues found while getting the containers working.

* Add draft query-ui container. Not yet ready for prime-time (needs an easier way of getting some
  data in to opensearch for easier use)
* Refactor containers so we have docker-base as a base with just the core of sycamore installed
  and then other containers built on top of that one.
* Switch to use --link for all the COPY commands to speed up builds
* Switch to use of Makefiles for installation -- worked very well for our cloud service code
* Make a way that dependent containers can check for consistent code
* Install all of the sycamore sub-packages. I was getting very weird behavior in installation
  in different directories with a shared venv -- stuff kept getting uninstalled that wasn't
  recognized. Since there were a lot of these, wrote a script to make the commands.
* Add poetry groups for all the sycamore optional bits to make them easier to install.
* Write script to do poetry installs in the most efficient way possible, taking advantage
  of caching wherever possible.  Record everything installed so that the jupyter container
  can auto-reinstall those things when switching to a github checkout source.
* Partially deprecate the importer
* Rename jupter fixuser.py to entrypoint.py since it's doing new stuff
* Fix several of the install commands to use the right poetry cache.
* Add support for the jupyter container to map in a sycamore git checkout and properly
  install
* Add hack that runs the raw stuff outside of the query ui. Otherwise the query ui is killing
  ray resulting in failures.  Add sycmoare_ray_init function that can connect sycamore ray to an
  external ray.
* Add support for a local disk cache to the query ui. Shared S3 cache doesn't work for non-Aryn
  engineers. Added a function that picks the right type of cache to use based on the specified
  path.
* Add support for ntsb_loader to use opensearch not on localhost:9200 or assuming it's at
  opensearch:9200 to allow for opensearch containers not running on the default port.
* Add an only materialized version of the ntsb loader. Something in the script needs more stuff
  to run even though it won't happen to execute.

mdwelsh reviewed (commented):
So, I know we discussed this, but I'd like to ask you to temporarily remove the cache.db file from this PR and we can discuss how to do this properly when you're back.

I tried to rewrite your README in apps/query-ui/cache_dir about how to build the file, and realized there are some issues here.

The main concern is that there's no easy way to rebuild the cache file without doing a bunch of gnarly steps. I guess you need to delete the file, re-run the query UI with the right env var set, and run specific queries with a specific LLM with a specific dataset loaded into a specific index. That is, it's not reproducible in a reliable and consistent way.

The problem is that "stale" cached LLM return values frozen in a file that we can't update easily limits our ability to rev many aspects of the system. It prevents us from updating the query planner, operator set, LLM parameters, etc. since those "frozen" queries stay stuck in the cache and do not get the benefit of those updates. If there were a simple, repeatable way to rebuild the cache, I would not have this concern, but currently there is not.

If all you care about is seeing a result fast for a couple of queries, then we should just hardcode the result of those queries in the Python code, since that is the moral equivalent of what the contents of this specific cache file is doing.

I'm also unconvinced that the small benefit of having a couple of simple test queries run magically fast outweighs the downsides described above. So, if this is mainly useful for Eric when testing things himself, my feeling is we shouldn't yet commit this cache to the git repo. Eric can still have his own local cache file that he uses for testing - nothing prevents you from doing that - but let's figure out how to do this right.

HenryL27 reviewed (commented):
couple questions/nits but overall docker side looks good. I'll let you and @mdwelsh figure out what to do about the question cache

baitsguy commented:
> So, I know we discussed this, but I'd like to ask you to temporarily remove the cache.db file from this PR and we can discuss how to do this properly when you're back.
> 
> I tried to rewrite your README in apps/query-ui/cache_dir about how to build the file, and realized there are some issues here.
> 
> The main concern is that there's no easy way to rebuild the cache file without doing a bunch of gnarly steps. I guess you need to delete the file, re-run the query UI with the right env var set, and run specific queries with a specific LLM with a specific dataset loaded into a specific index. That is, it's not reproducible in a reliable and consistent way.
> 
> The problem is that "stale" cached LLM return values frozen in a file that we can't update easily limits our ability to rev many aspects of the system. It prevents us from updating the query planner, operator set, LLM parameters, etc. since those "frozen" queries stay stuck in the cache and do not get the benefit of those updates. If there were a simple, repeatable way to rebuild the cache, I would not have this concern, but currently there is not.
> 
> If all you care about is seeing a result fast for a couple of queries, then we should just hardcode the result of those queries in the Python code, since that is the moral equivalent of what the contents of this specific cache file is doing.
> 
> I'm also unconvinced that the small benefit of having a couple of simple test queries run magically fast outweighs the downsides described above. So, if this is mainly useful for Eric when testing things himself, my feeling is we shouldn't yet commit this cache to the git repo. Eric can still have his own local cache file that he uses for testing - nothing prevents you from doing that - but let's figure out how to do this right.

Should we not just have a cache invalidate option? the ui can just say use cache, override cache, don't use cache. Doesn't seem unreasonable to me under an advanced drop down

eric-anderson commented:
> [ Remove the cache.db file ]

Done. I don't understand why you're ok with the s3 cache being checked in by default and not this one, but I moved the file to s3 and added instructions on how to apply it, so it's more awkward to use, but doesn't have the cache file committed into git.

eric-anderson commented on apps/docker-base/Makefile.docker-base:
```diff
@@ -0,0 +1,33 @@
+	test "$(GIT_COMMIT)" != "unknown"
+	touch .git.commit.$(GIT_COMMIT)
+
+check-version-compatibility:
+	test "$(GIT_COMMIT)" != ""
+	test "$(GIT_COMMIT)" != "unknown"
+	ls .git.commit.*
+	test -f .git.commit.$(GIT_COMMIT)
```
Done.

eric-anderson commented on apps/docker-base/poetry-install.sh:
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+set -e
+set -x
+if [[ ! -d /app/.venv ]]; then
+    echo "Missing /app/.venv; assuming about to do noop-poetry install"
```
There used to be code that installed a noop poetry environment back when i was trying to not have all the pyproject.toml files in the image. I couldn't make it work, but forgot to update the comment. Thanks.

eric-anderson commented on apps/docker-base/poetry-install.sh:
```diff
@@ -0,0 +1,37 @@
+fi
+
+if [[ $# = 0 ]]; then
+    echo "Usage: $0 <poetry install packages>"
```
Yes, although all optional. Updated to `poetry-install.sh [group1] [group2] [group3]`

eric-anderson commented on apps/docker-base/Makefile.docker-base:
```diff
@@ -0,0 +1,33 @@
+
+apt-install:
+	DEBIAN_FRONTEND=noninteractive apt update
+	# gcc and python3-dev needed on arm for guidance
```
On arm, yes, on x86 no.  I had initially removed these when I rewrote to the makefiles because I couldn't figure out why they were needed.  Then I built the arm images and it failed.

eric-anderson commented on apps/jupyter/bind_dir/setup.sh:
```diff
@@ -0,0 +1,39 @@
+        maybelink ../sycamore.git/$i /app/work/$i
+    done
+    # Reinstall everything that's been installed
+    for i in $(cat /app/.poetry.install); do
+        ./poetry-install.sh $i
+    done
```
Maybe. I'd rather keep it the way it is because then it works exactly the same as when it was originally run.

eric-anderson commented on apps/query-ui/run-query-ui.sh:
```diff
@@ -0,0 +1,39 @@
+while [[ "${OPENSEARCH_HOST}" == "" ]]; do
+    i=$(expr $i + 1)
+    if [[ $i -gt 120 ]]; then
+        echo "Waiting 120 seconds to find opensearch. Giving up."
```
Yes, yes there should.

eric-anderson commented on apps/query-ui/run-query-ui.sh:
```diff
@@ -0,0 +1,39 @@
+[[ "${QUERY_CACHE}" == "" ]] && export QUERY_CACHE=/app/work/cache_dir
+if [[ "${QUERY_CACHE}" == s3://* ]]; then
+    # e.g. s3://aryn-temp/llm_cache/luna/ntsb
+    :
```
It's the way shell writes the python "pass"

eric-anderson commented on apps/query-ui/run-query-ui.sh:
```diff
@@ -0,0 +1,39 @@
+
+echo "Inferred opensearch host ${OPENSEARCH_HOST}"
+    
+poetry run python apps/query-ui/external_ray_wrapper.py
```
I have something that works, so I'd rather not change it. It might work, but I can't tell -- ray wasn't finding the cluster I started in the parent until i did that magic. I left a note, but it will all be moot if Matt/Vinayak can figure out why streamlit is killing ray.

eric-anderson commented on compose.yaml:
```diff
@@ -90,6 +75,35 @@ services:
+    command: /bin/sh -c 'echo "starting chown..." && chown -R 1000:1000 /tmp/crawl_data && echo "chown successful"'
+
+  query-ui:
+    # ERIC: look into https://www.joyfulbikeshedding.com/blog/2021-03-15-docker-and-the-host-filesystem-owner-matching-problem.html
```
Leaving it is intentional it should just be TODO.

eric-anderson commented on compose.yaml:
```diff
@@ -122,20 +136,27 @@ services:
+    profiles: ['deprecated_importer']
+    image: arynai/sycamore-importer:${SYCAMORE_VERSION}
+    ports:
+      - 127.0.0.1:${RAY_CONSOLE_PORT}:8265 # note, not working right now
```
Maybe. I'm not going to try to figure that out for now.  Right now we like launching ray under our code rather than externally.
